### DISCUSSION TITLE
errorMessage: Unable to determine handler from trigger event

### DISCUSSION THREAD

**Discussion by VictorBac** [Kludex/mangum#218] [Q&A]
Hey guys,

I am trying to build d coker iamge to deploy on aws lambda following the doc and:
https://docs.aws.amazon.com/lambda/latest/dg/images-create.html#images-create-from-base
https://docs.aws.amazon.com/lambda/latest/dg/python-image.html#python-image-base

My Dockerfile:
```
FROM public.ecr.aws/lambda/python:3.8

# Copy function code
COPY app.py ${LAMBDA_TASK_ROOT}

# Install the function's dependencies using file requirements.txt
# from your project folder.

COPY requirements.txt  .
RUN  pip3 install -r requirements.txt --target "${LAMBDA_TASK_ROOT}"

# Set the CMD to your handler (could also be done as a parameter override outside of the Dockerfile)
CMD [ "app.handler" ]
``` 

requirement.txt:
``` 
fastapi == 0.71.*
uvicorn[standard]
mangum == 0.12
```

app.py:
``` 
from fastapi import FastAPI
from mangum import Mangum

app = FastAPI()


@app.get("/")
def read_root():
    return {"Hello": "World"}


@app.get("/items/{item_id}")
def read_item(item_id: int, q: str = None):
    return {"item_id": item_id, "q": q}

handler = Mangum(app, spec_version=2)
```

then:
```
docker build -t hello-world .
docker run -p 9000:8080 hello-world
```

But when I try to query locally (following the doc):
```
$ curl -XPOST "http://localhost:9000/2015-03-31/functions/function/invocations" -d '{}'
{
  "errorMessage": "Unable to determine handler from trigger event",
  "errorType": "TypeError",
  "stackTrace": [
    "  File \"/var/task/mangum/adapter.py\", line 86, in __call__\n    handler = AbstractHandler.from_trigger(\n",
    "  File \"/var/task/mangum/handlers/abstract_handler.py\", line 113, in from_trigger\n    raise TypeError(\"Unable to determine handler from trigger event\")\n"
  ]
}
```

I tried using (found in some discussions)
```
handler = Mangum(app, spec_version=2)
```

and even from another discussion:
``` 
import uvicorn
from fastapi import FastAPI
from mangum import Mangum
# TODO: Add this line
from starlette.middleware.cors import CORSMiddleware

app = FastAPI(title="FastAPI Mangum Example", version='1.0.0')

# TODO: Add these lines
app.add_middleware(
    CORSMiddleware,
    allow_origins='*',
    allow_credentials=False,
    allow_methods=["GET", "POST", "OPTIONS"],
    allow_headers=["x-apigateway-header", "Content-Type", "X-Amz-Date"],
)

handler = Mangum(app)


@app.get('/', name='Hello World', tags=['Hello'])
def hello_world():
    return {"Hello": "Python"}


if __name__ == '__main__':
    uvicorn.run(app)
```

it never works. (same error when deployed to aws lamda and not run locally)

I tried to use previous version but also had issue...
does anyone has a working example for this ?

**Comment by ddppddpp** [comment]
Please try using mangum 0.11 in your requirements. This fixed it for me. Not sure if it is an issue with AWS Lambda or mangum.

**Comment by jordaneremieff** [comment]
@VictorBac the problem appears to be that the local invocation does not include an event payload, so the Mangum handler is unable to identify the service invoking the function. Mangum uses the event data to determine which keys to map to the `scope` for the ASGI cycle and how to format the response.

Not sure what service you are using for your deployed environment, but you'll need a payload format like the one displayed in the [docs](https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-develop-integrations-lambda.html) for API Gateway requests.

**Reply by AlphonsG** [reply]
> @VictorBac the problem appears to be that the local invocation does not include an event payload, so the Mangum handler is unable to identify the service invoking the function. Mangum uses the event data to determine which keys to map to the `scope` for the ASGI cycle and how to format the response.
> 
> Not sure what service you are using for your deployed environment, but you'll need a payload format like the one displayed in the [docs](https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-develop-integrations-lambda.html) for API Gateway requests.

Run this when testing locally:

 curl -XPOST "http://localhost:8080/2015-03-31/functions/function/invocations/predict" -d '{"resource": "/", "path": "/api/v1/test/", "httpMethod": "GET", "requestContext": {}, "multiValueQueryStringParameters": null}'

**Comment by zsiegel92** [comment]
@AlphonsG your suggestion is correct!

If anyone wants to write Python tests to hit your endpoints, try this:

**`app.py`**:

```python
from fastapi import FastAPI
from mangum import Mangum

app = FastAPI()
handler = Mangum(app)


@app.get("/")
async def root():
    return {"message": "Hello World!"}
```

**`test_api.py`**:


```python
import json
import requests


def test_endpoint(endpoint: str):
    url = 'http://localhost:9000/2015-03-31/functions/function/invocations'
    response = requests.get(
        url,
        json={
            "resource": endpoint,
            "path": endpoint,
            "httpMethod": "GET",
            "requestContext": {},
            "multiValueQueryStringParameters": None,
        },
    )
    data = json.loads(response.json()['body'])
    return data


if __name__ == "__main__":
    response = test_endpoint("/")
    assert response['message'] == "Hello World!" # passes
```

**Reply by zsiegel92** [reply]
To run the Docker container and forward the internal handler port `8080` to `localhost:9000` as I did, use:

```sh
docker run --platform linux/amd64 --name CONTAINERNAME -p 9000:8080 IMAGENAME
```
where `CONTAINERNAME` and `IMAGENAME` should be replaced.
